### PR TITLE
Copy objects in federation test helper

### PR DESCRIPTION
**Pull request by mwielgus**
Push clean copies of passed objects in test helper in federation controllers tests.

Fixes: #33667

cc: @gmarek @quinton-hoole @kubernetes/sig-cluster-federation

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33769)

### CODE REVIEW

**Comment by gmarek** [comment]
```
=== RUN   TestIngressController
==================
WARNING: DATA RACE
Read by goroutine 20:
  reflect.Value.IsNil()
      /usr/local/go/src/reflect/value.go:970 +0x85
  reflect.deepValueEqual()
      /usr/local/go/src/reflect/deepequal.go:103 +0x98f
  reflect.deepValueEqual()
      /usr/local/go/src/reflect/deepequal.go:97 +0x14c5
  reflect.deepValueEqual()
      /usr/local/go/src/reflect/deepequal.go:97 +0x14c5
  reflect.deepValueEqual()
      /usr/local/go/src/reflect/deepequal.go:94 +0x1044
  reflect.DeepEqual()
      /usr/local/go/src/reflect/deepequal.go:185 +0x263
  k8s.io/kubernetes/federation/pkg/federation-controller/util.NewTriggerOnAllChanges.func3()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/federation/pkg/federation-controller/util/handlers.go:42 +0xa7
  k8s.io/kubernetes/pkg/client/cache.ResourceEventHandlerFuncs.OnUpdate()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/cache/controller.go:175 +0x5b
  k8s.io/kubernetes/pkg/client/cache.(*ResourceEventHandlerFuncs).OnUpdate()
      <autogenerated>:51 +0xed
  k8s.io/kubernetes/pkg/client/cache.NewInformer.func1()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/cache/controller.go:241 +0x379
  k8s.io/kubernetes/pkg/client/cache.(*DeltaFIFO).Pop()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/cache/delta_fifo.go:420 +0x3f5
  k8s.io/kubernetes/pkg/client/cache.(*Controller).processLoop()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/cache/controller.go:126 +0x78
  k8s.io/kubernetes/pkg/client/cache.(*Controller).(k8s.io/kubernetes/pkg/client/cache.processLoop)-fm()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/cache/controller.go:102 +0x2d
  k8s.io/kubernetes/pkg/util/wait.JitterUntil.func1()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:84 +0x2a
  k8s.io/kubernetes/pkg/util/wait.JitterUntil()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:85 +0xc9
  k8s.io/kubernetes/pkg/util/wait.Until()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:47 +0x50
  k8s.io/kubernetes/pkg/client/cache.(*Controller).Run()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/cache/controller.go:102 +0x25d

Previous write by goroutine 17:
  k8s.io/kubernetes/federation/pkg/federation-controller/ingress.TestIngressController()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/federation/pkg/federation-controller/ingress/ingress_controller_test.go:143 +0x35db
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:473 +0xdc

Goroutine 20 (running) created at:
  k8s.io/kubernetes/federation/pkg/federation-controller/ingress.(*IngressController).Run()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/federation/pkg/federation-controller/ingress/ingress_controller.go:278 +0xb0
  k8s.io/kubernetes/federation/pkg/federation-controller/ingress.TestIngressController()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/federation/pkg/federation-controller/ingress/ingress_controller_test.go:95 +0x1c4b
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:473 +0xdc

Goroutine 17 (running) created at:
  testing.RunTests()
      /usr/local/go/src/testing/testing.go:582 +0xae2
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:515 +0x11d
  main.main()
      k8s.io/kubernetes/federation/pkg/federation-controller/ingress/_test/_testmain.go:54 +0x210
```

**Comment by mwielgus** [comment]
Added extra copies before returning data to the test via channel. Should be better now.

File: federation/pkg/federation-controller/util/test/test_helper.go [hunk 1]
```diff
@@ -48,14 +49,22 @@ func (wd *WatcherDispatcher) register(watcher *watch.FakeWatcher) {
 	}
 }
 
+func copy(obj runtime.Object) runtime.Object {
+	obj, err := api.Scheme.Copy(obj)
```
Comment (quinton-hoole) [inline federation/pkg/federation-controller/util/test/test_helper.go]: Drive-by comment: Don't you need a deep, not a shallow copy here?

Comment (mwielgus) [inline federation/pkg/federation-controller/util/test/test_helper.go]: Deep copy.

**Comment by mwielgus** [comment]
@k8s-bot test this issue: #33672

**Comment by mwielgus** [comment]
@k8s-bot gke e2e test this issue: #33672

File: federation/pkg/federation-controller/util/test/test_helper.go (hunk 1 above)
Comment (gmarek) [inline federation/pkg/federation-controller/util/test/test_helper.go]: I believe that copy is a shallow copy. DeepCopy is Scheme.DeepCopy.

Comment (mwielgus) [inline federation/pkg/federation-controller/util/test/test_helper.go]: Wow :). What's the point of having a shallow copy?
Fixed.

**Comment by gmarek** [comment]
Thanks for doing it!
